### PR TITLE
Make run button functional

### PR DESCRIPTION
**Pull request by david-mears-2**
* Add a functional submit button and all the related gubbins: api endpoint and handler, tests.
* Use a pinia store for largeScreen

You can test this locally by seeing if clicking the run button takes you to a page showing you the dummy run id.

### CODE REVIEW

File: components/ParameterForm.vue
Comment (absternator) [inline components/ParameterForm.vue:139]: should defs use body here... your code shouldn't change to accommodate the tests...

File: components/ParameterForm.vue
Comment (absternator) [inline components/ParameterForm.vue:157]: should move this at top with all variable definitions... nice to keep nicely separated with methods

File: components/ParameterForm.vue
Comment (absternator) [inline components/ParameterForm.vue:132]: when does this happen?

File: components/ParameterForm.vue
Comment (absternator) [inline components/ParameterForm.vue:143]: display this error back to user

File: components/ParameterForm.vue
Comment (absternator) [inline components/ParameterForm.vue:148]:
```suggestion
   await navigateTo(`/scenarios/${runId}`)
```

unless reason for assignment before?

File: components/ParameterForm.vue
Comment (absternator) [inline components/ParameterForm.vue:2]: is this for playwright again? i feel its causing another render now... probs best to take the L in the playwright test rather than decrease app performance and increase bloat

File: components/ParameterForm.vue
Comment (absternator) [inline components/ParameterForm.vue:74]: a loading icon could be nice also whilst `formSubmitting`

File: components/SideBar.vue
Comment (absternator) [inline components/SideBar.vue:9]: is data-testid neccesary? can it be tartged without this in tests? i am just against data-testid because its something users cant see and imaginary of sorts... and tests should mimic what users have access too

File: server/api/scenarios.post.ts
Comment (absternator) [inline server/api/scenarios.post.ts:8]: as per prev comment should be reading body not query params

File: server/api/scenarios.post.ts
Comment (absternator) [inline server/api/scenarios.post.ts:12]: dont need this comment

File: server/handlers/scenarios.ts
Comment (absternator) [inline server/handlers/scenarios.ts:8]: dont neeed comments again.. the code you wrote is clean and understandable

File: stores/appStore.ts
Comment (absternator) [inline stores/appStore.ts:17]: should this be `setScreenSizeToLarge` .. also function param should be descrptive .. eg. isLargeScreen rather than bool

**Review by absternator — COMMENTED** [review]
Nice work !!! looks clean and works great!!!!!! just left some minor comments

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:139]: I don't have a reason to prefer body over query; do you?

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:132]: When does !formData.value be undefined, you mean? It would only happen if the inputs were nullish, which shouldn't happen, but possibly could with some combination of form autofilling and hydration lifecycle race conditions? It's defensive coding.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:148]: This is for testing only, because I couldn't get a spy on navigateTo.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:2]: Another render?

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:74]: Yep, I put a spinner into the button. Not that it hangs around long enough for a human to see it!

File: components/SideBar.vue
Comment (david-mears-2) [reply components/SideBar.vue:9]: It's a handle for playwright to grab onto. Playwright has a limited range of locators, but it does also have 'getByRole' that we could use here - except that we have two things with a navigation role so it's not unique. Also, when the navbar is narrow, it has no text that we could have done getByText on.

File: stores/appStore.ts
Comment (david-mears-2) [reply stores/appStore.ts:17]: Good points.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:143]: Making a ticket

Comment (david-mears-2) [reply components/ParameterForm.vue:143]: Context - I only expect there to be errors here in development, since there is no known way for the form data to fail validation at the moment (you can't empty the input fields)

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:2]: Let me know if you manage to reproduce the re-rendering because I can't make it happen for me -- would be great to have steps to reproduce.

File: components/ParameterForm.vue
Comment (absternator) [reply components/ParameterForm.vue:139]: Semantically the use of query params is incorrect here... they are more for filtering, sorting / picking certain ids... when doing post request and sending data its semantically correct to use a body

File: components/ParameterForm.vue
Comment (absternator) [reply components/ParameterForm.vue:143]: sweet

File: components/ParameterForm.vue
Comment (absternator) [reply components/ParameterForm.vue:2]: nah you are good i just thought it would

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:74]: (maybe it will though when the subsequent page takes actual time to load)

File: server/api/scenarios.post.ts
Comment (david-mears-2) [reply server/api/scenarios.post.ts:12]: I've moved this comment - which was getting duplicated for each api endpoint - into a new .md "API design decisions.md"

File: stores/appStore.ts
Comment (david-mears-2) [reply stores/appStore.ts]: I have changed the function param to be called large, otherwise kept this the same

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:141]: I tried for some time to use the nuxt test utils to pass through a body, but there didn't seem to be a way.
To be clear, the R API is still receiving a body, it's only the web app that's taking a query parameter.

File: tests/unit/components/ParameterForm.spec.ts
Comment (EmmaLRussell) [inline tests/unit/components/ParameterForm.spec.ts:168]: Did you try stubGlobal or mockNuxtImport?
https://stackoverflow.com/questions/77827093/how-to-check-if-navigateto-was-called-in-nuxt-js

**Review by EmmaLRussell — COMMENTED** [review]
Great stuff, good to get the store in. 

Seems like some of the testing was a headache, which I'm probably adding to with some of these comments.. I think it would be really good to avoid some of these test-related-code-polluting additions if possible. Have made a couple of suggestions which you may have already tried. Happy to have a look at any of these if you're sick of bashing against them yourself - can't guarantee to get anywhere, as this is mostly all new to me too, but I think it's worth solving these now if we can, and can avoid proliferation.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:141]: I tried again today with a fresh brain and am making headway :)

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:2]: Luckily, this problem disappears [next PR](https://github.com/jameel-institute/daedalus-web-app/pull/27), because by then we can wait for version data as our signal that everything is mounted.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:154]: I had the comment in the test file - there are two other places it could be, here, and by the data attribute itself.

Comment (david-mears-2) [reply components/ParameterForm.vue:154]: I think our only option besides this is just not to test the form submission as part of component tests. That means we would rely only on e2e tests for the form submission. But since our e2e test doesn't mock the R API, we can't verify that the correct parameters were passed to it by the component.

File: stores/appStore.ts
Comment (david-mears-2) [reply stores/appStore.ts:16]: True. Previously this had more going on!

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:145]: I was imagining all of the scenario-related endpoints living under a single 'scenarios' namespace, rather than varying between 'scenario' and 'scenarios' depending on the use case. I believe in CRUD it is conventional to standardize on the plural form in route names.

File: components/SideBar.vue
Comment (david-mears-2) [reply components/SideBar.vue:9]: It happens that this is vestigial anyway.

File: layouts/default.vue
Comment (david-mears-2) [reply layouts/default.vue:31]: I went through an adventure since writing this, submitted a PR to clarify the docs of the nuxt pinia module, and eventually the maintainer let me know that in fact "you don't need to pass the pinia instance within setup"
https://github.com/vuejs/pinia/pull/2766

As such, the next PR removes this, here: https://github.com/jameel-institute/daedalus-web-app/pull/27/files#diff-433a9abd4ca70520f69fc064c160bb093918ce26dda6087a8bbfdbe2895d1944

File: tests/e2e/runAnalysis.spec.ts
Comment (david-mears-2) [reply tests/e2e/runAnalysis.spec.ts:20]: We can be confident that the values of the inputs are what we just selected. I think the test to do for the e2e would be to wait until the next page is implemented, and test that the displayed parameters are correct.

File: tests/integration/rApi.spec.ts
Comment (david-mears-2) [reply tests/integration/rApi.spec.ts]: My problem had been that I was trying to pass objects into the 'body' property that just got lost in the post, as it were. I eventually solved this by finding that I could wrap the parameters in a `FormData` object, and decompose that at the other end.

File: tests/unit/components/ParameterForm.spec.ts
Comment (david-mears-2) [reply tests/unit/components/ParameterForm.spec.ts:168]: Yes.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:7]: Removed the dependency on this from tests. In the component test, I think the test relying on this was redundant anyway.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue]: Removed in next PR.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue]: Removed in next PR.

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:154]: Well, until such time as it can test the result page, I suppose.

File: stores/appStore.ts
Comment (david-mears-2) [reply stores/appStore.ts]: Good to know.

File: components/ParameterForm.vue
Comment (EmmaLRussell) [reply components/ParameterForm.vue:143]: Yeah, it's a bit uneasy. Fair enough!

File: tests/e2e/runAnalysis.spec.ts
Comment (EmmaLRussell) [reply tests/e2e/runAnalysis.spec.ts]: Hang on, that doesn't look like the line (I thought) this was a comment for anymore, I'm pretty sure this was about the stringified JSON - or I thought it was!

Comment (david-mears-2) [reply tests/e2e/runAnalysis.spec.ts]: Still looks like it's commenting on the stringified JSON to me!

File: tests/unit/components/ParameterForm.spec.ts
Comment (EmmaLRussell) [reply tests/unit/components/ParameterForm.spec.ts:135]: OK, let's not hold up this branch, but it would be really good to find a way to mock global methods, and it seems really weird if the alleged mocking capacities of vitest/nuxt test utils aren't going to let us to that. Tbh if that proves impossible personally I'd rather not clutter up the code with extra props for the sake of unit tests, as long as the navigation etc gets tested in e2e tests. These components are going to be busy and confusing enough as it is, without a bunch of extra stuff that doesn't really need to be there.

File: components/ParameterForm.vue
Comment (EmmaLRussell) [inline components/ParameterForm.vue:140]: I'd assumed that we were just going to post JSON in the same format that the R API accepts and pass it straight through? You're saying that doesn't work (in the tests), that ordinary objects get turned into... what? Completely empty body?

It's not something to do with needing to set the content type is it? Does it default to POSTing form encoded data so it barfs on plain objects but accepts FormData? Hm, the ofetch docs suggest it should deal with that transparently and add the header itself, but it is supicious... Could try setting that header explicitly?

**Review by EmmaLRussell — COMMENTED** [review]
I think at some point we're going to have to work out how to POST JSON with Nuxt tests, as it really ought to be possible 😆 It's not terrible to post form values as form data but it would be nice to have the option not to.. 

See previous comment on the mocking navigate controversy - I'd rather take out the props that are just there for unit tests if it's really not possible to mock, and make sure the navigration is tested in e2e. I might have a fiddle around with the mocking, maybe I'll stumble on some other way of doing it... but that can go in another branch.

File: tests/e2e/runAnalysis.spec.ts
Comment (EmmaLRussell) [reply tests/e2e/runAnalysis.spec.ts]: Sorry I'm just blind as a :bat:

**Comment by EmmaLRussell** [comment]
> See previous comment on the mocking navigate controversy - I'd rather take out the props that are just there for unit tests if it's really not possible to mock, and make sure the navigration is tested in e2e. I might have a fiddle around with the mocking, maybe I'll stumble on some other way of doing it... but that can go in another branch.

This seemed to work for me: https://github.com/jameel-institute/daedalus-web-app/pull/29/files#

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:152]: Emma found this solution: https://github.com/jameel-institute/daedalus-web-app/pull/29

File: components/ParameterForm.vue
Comment (david-mears-2) [reply components/ParameterForm.vue:140]: I have found something that appears to be working - JSON stringifying the body (in the test). The oftech docs claimed that this happened automatically, but evidently not?

Comment (david-mears-2) [reply components/ParameterForm.vue]: See commit 6530321 for solution

Comment (david-mears-2) [reply components/ParameterForm.vue]: It seems that still it behaves differently under test than in real life, so I had to add in a check for whether the body was being received as an object (as in real life, and in e2e tests) or a JSON parsable string (as in the integration test). I hope this is tolerable, as the price for moving away from FormData.

Comment (EmmaLRussell) [reply components/ParameterForm.vue]: As discussed. maybe try adding header `Content-Type: application/json` when doing the POST.

**Review by EmmaLRussell — APPROVED** [review]
OK this one's cooked as far as I'm concerned!

**Pull request merged**
